src/charm.py

```diff
@@ -0,0 +1,315 @@
+            charm=self,
+            ports=[
+                ServicePort(name="prometheus-exporter", port=PROMETHEUS_PORT),
+                ServicePort(name="sbi", port=SBI_PORT),
+                ServicePort(name="ngapp", port=NGAPP_PORT, protocol="SCTP"),
+                ServicePort(name="sctp-grpc", port=SCTP_GRPC_PORT),
+            ],
+        )
+        self._amf_database = DatabaseRequires(
+            self, relation_name="amf-database", database_name=AMF_DATABASE_NAME
+        )
+        self._default_database = DatabaseRequires(
+            self, relation_name="default-database", database_name=DEFAULT_DATABASE_NAME
+        )
+
+        self.framework.observe(
+            self.on.default_database_relation_joined,
+            self._configure_amf,
+        )
+        self.framework.observe(
+            self.on.amf_database_relation_joined,
+            self._configure_amf,
+        )
+        self.framework.observe(self._default_database.on.database_created, self._configure_amf)
+        self.framework.observe(self._amf_database.on.database_created, self._configure_amf)
+        self.framework.observe(self.on.amf_pebble_ready, self._configure_amf)
+        self.framework.observe(self._nrf_requires.on.nrf_available, self._configure_amf)
+
+    def _configure_amf(
+        self,
+        event: EventBase,
+    ) -> None:
+        """Handle pebble ready event for AMF container.
+
+        Args:
+            event (PebbleReadyEvent, DatabaseCreatedEvent, NRFAvailableEvent): Juju event
+        """
+        if not self._amf_container.can_connect():
+            self.unit.status = MaintenanceStatus("Waiting for service to start")
+            event.defer()
+            return
+        for relation in ["fiveg_nrf", "amf-database", "default-database"]:
+            if not self._relation_created(relation):
+                self.unit.status = BlockedStatus(f"Waiting for {relation} relation")
+                return
+        if not self._default_database_is_available():
+            self.unit.status = WaitingStatus("Waiting for the default database to be available")
+            event.defer()
+            return
+        if not self._amf_database_is_available():
+            self.unit.status = WaitingStatus("Waiting for the amf database to be available")
+            event.defer()
+            return
+        if not self._get_default_database_info():
+            self.unit.status = WaitingStatus("Waiting for default database info to be available")
+            event.defer()
+            return
+        if not self._nrf_requires.nrf_url:
+            self.unit.status = WaitingStatus("Waiting for NRF data to be available")
+            event.defer()
+            return
+        if not self._amf_container.exists(path=CONFIG_DIR_PATH):
+            self.unit.status = WaitingStatus("Waiting for storage to be attached")
+            event.defer()
+            return
+        content = self._render_config_file(
```

Comment:
I'd still extract all the logic related to generating a config file to a separate function. It'd look cleaner ;)